### PR TITLE
8307100: Remove ReferentBasedDiscovery reference discovery policy

### PR DESCRIPTION
**Pull request by albertnetymk**
Mostly consisting of mechanic refactoring after replacing `RefDiscoveryPolicy == ...` with `true` or `false`.

Test: tier1-6

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8307101](https://bugs.openjdk.org/browse/JDK-8307101) to be approved

### Issues
 * [JDK-8307100](https://bugs.openjdk.org/browse/JDK-8307100): Remove ReferentBasedDiscovery reference discovery policy
 * [JDK-8307101](https://bugs.openjdk.org/browse/JDK-8307101): Remove ReferentBasedDiscovery reference discovery policy (**CSR**)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [26848ab2](https://git.openjdk.org/jdk/pull/13715/files/26848ab2c40941c432c903693c92a4d5d3f5d17c)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [26848ab2](https://git.openjdk.org/jdk/pull/13715/files/26848ab2c40941c432c903693c92a4d5d3f5d17c)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13715/head:pull/13715` \
`$ git checkout pull/13715`

Update a local copy of the PR: \
`$ git checkout pull/13715` \
`$ git pull https://git.openjdk.org/jdk.git pull/13715/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13715`

View PR using the GUI difftool: \
`$ git pr show -t 13715`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13715.diff">https://git.openjdk.org/jdk/pull/13715.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13715#issuecomment-1527489218)

### CODE REVIEW

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by dholmes-ora — APPROVED** [review]
The obsoletion of RefDiscoveryPolicy by always using refernce-based seems fine.

One query below.

Thanks.

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -523,6 +523,7 @@ static SpecialFlag const special_jvm_flags[] = {
   { "G1UsePreventiveGC",            JDK_Version::undefined(), JDK_Version::jdk(21), JDK_Version::jdk(22) },
   { "G1ConcRSLogCacheSize",         JDK_Version::undefined(), JDK_Version::jdk(21), JDK_Version::undefined() },
   { "G1ConcRSHotCardLimit",         JDK_Version::undefined(), JDK_Version::jdk(21), JDK_Version::undefined() },
+  { "RefDiscoveryPolicy",           JDK_Version::undefined(), JDK_Version::jdk(21), JDK_Version::undefined() },
```
[the remark this answers is not in the document]
Comment (albertnetymk) [reply src/hotspot/share/runtime/arguments.cpp:526]: Not really; mostly copying from its neighbor.

File: src/hotspot/share/gc/shared/referenceProcessor.cpp [hunk 2]
```diff
@@ -921,32 +918,16 @@ bool ReferenceProcessor::is_subject_to_discovery(oop const obj) const {
...
+// Reference discovery policy:
+//   if the reference object is not in the "originating generation"
+//   (or part of the heap being collected, indicated by our "span")
+//   we don't treat it specially (i.e. we scan it as we would
+//   a normal oop, treating its references as strong references).
+//   This means that references can't be discovered unless their
+//   referent is also in the same span. This is the simplest,
+//   most "local" and most conservative approach, albeit one
+//   that may cause weak references to be enqueued least promptly.
+//   We call this choice the "ReferenceBasedDiscovery" policy.
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/referenceProcessor.cpp:930]: I would remove the last sentence; after this change "ReferenceBasedDiscovery" is never referenced anywhere useful, so mentioning it here does not seem to have a purpose. Not sure if it makes sense to keep indenting the paragraph explaining it either.

**Review by tschatzl — APPROVED** [review]
lgtm.

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**